File: ats-service/app/analysis/sections.py
"""Resume section detection and structural analysis."""
import re
from typing import Dict, Any, List, Tuple

SECTION_PATTERNS = {
    "experience": [
        r"^(work\s+)?experience",
        r"^professional\s+experience",
        r"^employment(\s+history)?",
        r"^work\s+history",
        r"^relevant\s+experience",
        r"^career\s+history",
        r"^internships?"
    ],
    "education": [
        r"^education(al)?(\s+background)?",
        r"^academic(\s+background|\s+history)?",
        r"^academics?",
        r"^qualifications?",
        r"^degrees?"
    ],
    "skills": [
        r"^(technical\s+|core\s+|key\s+)?skills",
        r"^technologies(\s+and\s+tools)?",
        r"^tools\s+(&|and)\s+technologies",
        r"^core\s+competencies",
        r"^technical\s+proficiencies?",
        r"^skillset"
    ],
    "projects": [
        r"^(key\s+|personal\s+|academic\s+|selected\s+|featured\s+)?projects",
        r"^portfolio\s+projects",
        r"^notable\s+work"
    ],
    "certifications": [
        r"^(licenses\s+(&|and)\s+)?certifications?",
        r"^certificates?",
        r"^accreditations?",
        r"^courses\s+(&|and)\s+certifications"
    ],
    "summary": [
        r"^(professional\s+|career\s+|executive\s+)?summary",
        r"^profile",
        r"^about\s+me",
        r"^(career\s+)?objective"
    ]
}

STANDARD_ORDER = ["summary", "skills", "experience", "projects", "education", "certifications"]

class SectionAnalyzer:
    """Detects, maps, and evaluates section headers in resume text."""

    @staticmethod
    def identify_heading(line: str) -> Tuple[bool, str]:
        """Checks if a given line is a section heading and returns (is_heading, canonical_name)."""
        clean = line.strip().lower()
        clean = re.sub(r"[:\-_|•\*\#]+", "", clean).strip()

        # Headings are typically short: 1 to 4 words, < 40 characters
        if not clean or len(clean) > 45 or len(clean.split()) > 5:
            return False, ""

        for canonical, patterns in SECTION_PATTERNS.items():
            for pat in patterns:
                if re.match(pat, clean):
                    return True, canonical

        return False, ""

    @classmethod
    def segment_resume(cls, full_text: str) -> Dict[str, Any]:
        """Splits resume text into detected sections and content."""
        lines = [l.strip() for l in full_text.splitlines() if l.strip()]
        sections: Dict[str, List[str]] = {}
        detected_order: List[str] = []
        current_section = "header"
        sections[current_section] = []

        unrecognized_headings: List[str] = []

        for line in lines:
            is_heading, canonical = cls.identify_heading(line)
            if is_heading:
                current_section = canonical
                if canonical not in sections:
                    sections[canonical] = []
                if canonical not in detected_order:
                    detected_order.append(canonical)
            else:
                # Check for possible unrecognized all-caps heading
                if (line.isupper() or line.istitle()) and len(line) < 30 and len(line.split()) <= 3:
                    # Might be an unusual heading
                    if line.lower() not in ["resume", "curriculum vitae", "cv", "phone", "email"]:
                        unrecognized_headings.append(line)
                sections[current_section].append(line)

        # Standard required sections: experience, education, skills
        present_sections = [s for s in detected_order if s in SECTION_PATTERNS]
        all_canonical = list(SECTION_PATTERNS.keys())
        missing_sections = [s for s in all_canonical if s not in present_sections]

        # Evaluate ordering logic
        # Standard: Contact/Header -> Summary -> (Experience or Skills) -> Projects -> Education
        is_order_good = True
        order_explanation = "Standard, ATS-friendly section sequence detected."

        if "experience" in detected_order and "education" in detected_order:
            exp_idx = detected_order.index("experience")
            edu_idx = detected_order.index("education")
            # For experienced professionals, experience before education is preferred
            # But either way is fine; if education is placed between random bullet points, that's bad

        if "experience" not in present_sections:
            is_order_good = False
            order_explanation = "Work Experience section was not clearly recognized."

        return {
            "sections": {k: "\n".join(v) for k, v in sections.items()},
            "present_sections": present_sections,
            "missing_sections": missing_sections,
            "detected_order": detected_order,
            "is_order_good": is_order_good,
            "order_explanation": order_explanation,
            "unrecognized_headings": list(set(unrecognized_headings))[:5]
        }

File: src/lib/ats/sections.js
/**
 * Resume section detection and structural analysis.
 * Identifies standard ATS sections and evaluates section hierarchy.
 */

export const SECTION_PATTERNS = {
  experience: [
    /^(work\s+)?experience/i,
    /^professional\s+experience/i,
    /^employment(\s+history)?/i,
    /^work\s+history/i,
    /^relevant\s+experience/i,
    /^career\s+history/i,
    /^internships?/i
  ],
  education: [
    /^education(al)?(\s+background)?/i,
    /^academic(\s+background|\s+history)?/i,
    /^academics?/i,
    /^qualifications?/i,
    /^degrees?/i
  ],
  skills: [
    /^(technical\s+|core\s+|key\s+)?skills/i,
    /^technologies(\s+and\s+tools)?/i,
    /^tools\s+(&|and)\s+technologies/i,
    /^core\s+competencies/i,
    /^technical\s+proficiencies?/i,
    /^skillset/i
  ],
  projects: [
    /^(key\s+|personal\s+|academic\s+|selected\s+|featured\s+)?projects/i,
    /^portfolio\s+projects/i,
    /^notable\s+work/i
  ],
  certifications: [
    /^(licenses\s+(&|and)\s+)?certifications?/i,
    /^certificates?/i,
    /^accreditations?/i,
    /^courses\s+(&|and)\s+certifications/i
  ],
  summary: [
    /^(professional\s+|career\s+|executive\s+)?summary/i,
    /^profile/i,
    /^about\s+me/i,
    /^(career\s+)?objective/i
  ]
};

export const STANDARD_CANONICALS = ['summary', 'skills', 'experience', 'projects', 'education', 'certifications'];

export function identifyHeading(line) {
  if (!line) return { isHeading: false, canonical: '' };
  const clean = line.replace(/[:\-_|•*#]+/g, ' ').trim().toLowerCase();
  const words = clean.split(/\s+/).filter(Boolean);

  if (words.length === 0 || words.length > 5 || clean.length > 45) {
    return { isHeading: false, canonical: '' };
  }

  for (const [canonical, patterns] of Object.entries(SECTION_PATTERNS)) {
    for (const pat of patterns) {
      if (pat.test(clean)) {
        return { isHeading: true, canonical };
      }
    }
  }

  return { isHeading: false, canonical: '' };
}

export function segmentResume(fullText) {
  const lines = fullText.split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
  const sections = { header: [] };
  const detectedOrder = [];
  let currentSection = 'header';

  for (const line of lines) {
    const { isHeading, canonical } = identifyHeading(line);
    if (isHeading) {
      currentSection = canonical;
      if (!sections[currentSection]) {
        sections[currentSection] = [];
      }
      if (!detectedOrder.includes(canonical)) {
        detectedOrder.push(canonical);
      }
    } else {
      if (!sections[currentSection]) {
        sections[currentSection] = [];
      }
      sections[currentSection].push(line);
    }
  }

  // Convert array of lines to joined string for each section
  const sectionTextMap = {};
  for (const [sec, secLines] of Object.entries(sections)) {
    sectionTextMap[sec] = secLines.join('\n');
  }

  const present_sections = detectedOrder.filter((s) => s in SECTION_PATTERNS);
  const missing_sections = STANDARD_CANONICALS.filter((s) => !present_sections.includes(s));

  // Determine section order health
  let is_order_good = true;
  // If summary appears, it should be near top
  const summaryIdx = detectedOrder.indexOf('summary');
  if (summaryIdx > 1) {
    is_order_good = false;
  }
  // Experience and Education should precede Certifications
  const expIdx = detectedOrder.indexOf('experience');
  const certIdx = detectedOrder.indexOf('certifications');
  if (expIdx !== -1 && certIdx !== -1 && certIdx < expIdx) {
    is_order_good = false;
  }

  return {
    sections: sectionTextMap,
    present_sections,
    missing_sections,
    detected_order: detectedOrder,
    is_order_good
  };
}
